### PR TITLE
Generate documentation from TASTY in sbt plugin

### PR DESCRIPTION
**Pull request by Duhemm**

### CODE REVIEW

**Comment by Duhemm** [comment]
@nicolasstucki I've rebased and adapted it to the changes with `-from-tasty`. If you want to try it, I suggest trying on a small project (dotty-language-server or dotty-doc-bootstrapped), or maybe running sbt with a very large heap?

File: compiler/src/dotty/tools/dotc/fromtasty/TASTYCompilationUnit.scala [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+import dotty.tools.dotc.core.Contexts.Context
 import dotty.tools.dotc.util.NoSource
+import dotty.tools.io.Path
 
 class TASTYCompilationUnit(val className: String) extends CompilationUnit(NoSource) {
   override def toString: String = s"class file $className"
 }
+object TASTYCompilationUnit {
+  def apply(className: String)(implicit ctx: Context): Option[TASTYCompilationUnit] = {
+    if (Path(className).exists) {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/fromtasty/TASTYCompilationUnit.scala]: `File(className)`

File: compiler/src/dotty/tools/dotc/fromtasty/TASTYCompilationUnit.scala [hunk 2]
```diff
@@ -1,8 +1,20 @@
...
+      ctx.inform(s"Ignoring $className: cannot create a `TASTYCompilationUnit` for a source file.")
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/fromtasty/TASTYCompilationUnit.scala]: I am not sure I understand. What if it is a TASTY file

Comment (Duhemm) [reply compiler/src/dotty/tools/dotc/fromtasty/TASTYCompilationUnit.scala]: Then it'll be excluded, but at this point the classname should have been extracted from it.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/fromtasty/TASTYCompilationUnit.scala]: Add some documentation

**Comment by nicolasstucki** [comment]
Is there a way to add a test for this?

**Comment by Duhemm** [comment]
We have one scripted test that runs the doc. The test is run nightly I think (@allanrenucci can confirm). 

https://github.com/lampepfl/dotty/blob/f66225c4cd85683316954365653e27cd21426c23/sbt-dotty/sbt-test/sbt-dotty/example-project/test#L1-L5

You can run it with `sbt-dotty/scripted sbt-dotty/example-project`.

File: sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala [hunk 3]
```diff
@@ -216,9 +216,19 @@ object DottyPlugin extends AutoPlugin {
...
       }
-    )
+    ) ++ inConfig(Compile)(docSettings) ++ inConfig(Test)(docSettings)
   }
 
+  private val docSettings = inTask(doc)(Seq(
+    sources := {
+      val _ = compile.value // Ensure that everything is compiled, so TASTy is available.
+      val prev = sources.value
+      val tastyFiles = (classDirectory.value ** "*tasty").get.map(_.getAbsoluteFile)
```
Comment (nicolasstucki) [inline sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala]: Maybe it should be `"*.tasty"`

**Pull request merged**